amida-tech/indaba-client
# INBA 707- Logout and return to referred fail point.

I'm not a giant fan of how I'm doing this, but it works and this solution probably isn't far off the mark. We can talk about improvements now or after MVP. 

#### What does this PR do?
When your token expires, your next request call will snag the 401 and return you to the login screen. When you fill out your username and password again, you will return to the screen that you expired on.

#### Related JIRA tickets:
https://jira.amida-tech.com/browse/INBA-707

#### How should this be manually tested?
Login as any user. 

Go to your console, applications. Modify or delete the `indaba-auth` token, but not the `indaba-realm` because that's technically still valid. Click on... almost anything that redirects or calls the backend service (as in, probably not any filters or modals). You should be returned to the login screen.

Fill in the username and password, then login. You should be pushed right back to where you were going. 

I've checked, and although the messaging api services handle their errors slightly differently, it still seems to work just fine. 

#### Background/Context
What's going on is that I went back on all the errors handing for the common actions and had them return err (for error status code) and errorMessage (the custom, translated error message meant for toasts). A middleware function now sits atop of all actions and checks for that err status of 401, and if found will immediately return the user to the login page, storing the referred link in the login redux state. That'll be referenced and pushed to during login and then removed.

@Perry5, `decodeResponse` was throwing an error under `requests.js` if it was unauthorized. The body text it was getting was in fact just "Unauthorized" so it was throwing crap when it tried to `JSON.parse()`, unknown token "U." We could change Greyscale to return a body that reads "Unauthorized" instead.

#### Screenshots (if appropriate):
